tests/unit/pubsublocal.tcl

```diff
@@ -0,0 +1,221 @@
+        assert_equal 2 [r publishlocal chan1 hello]
+        assert_equal {message chan1 hello} [$rd1 read]
+        assert_equal {message chan1 hello} [$rd2 read]
+
+        # clean up clients
+        $rd1 close
+        $rd2 close
+    }
+
+    test "PUBLISH/SUBSCRIBE after UNSUBSCRIBE without arguments" {
+        set rd1 [redis_deferring_client]
+        assert_equal {1 2 3} [subscribelocal $rd1 {chan1 chan2 chan3}]
+        unsubscribelocal $rd1
+        assert_equal 0 [r publishlocal chan1 hello]
+        assert_equal 0 [r publishlocal chan2 hello]
+        assert_equal 0 [r publishlocal chan3 hello]
+
+        # clean up clients
+        $rd1 close
+    }
+
+    test "SUBSCRIBE to one channel more than once" {
+        set rd1 [redis_deferring_client]
+        assert_equal {1 1 1} [subscribelocal $rd1 {chan1 chan1 chan1}]
+        assert_equal 1 [r publishlocal chan1 hello]
+        assert_equal {message chan1 hello} [$rd1 read]
+
+        # clean up clients
+        $rd1 close
+    }
+
+    test "UNSUBSCRIBE from non-subscribed channels" {
+        set rd1 [redis_deferring_client]
+        assert_equal {0 0 0} [unsubscribelocal $rd1 {foo bar quux}]
+
+        # clean up clients
+        $rd1 close
+    }
+
+    test "PUBSUB command basics" {
+        r pubsub local numsub abc def
+    } {abc 0 def 0}
+
+    test "PUBLISHLOCAL/SUBSCRIBELOCAL with two clients" {
+        set rd1 [redis_deferring_client]
+        set rd2 [redis_deferring_client]
+
+        assert_equal {1} [subscribelocal $rd1 {chan1}]
+        assert_equal {1} [subscribelocal $rd2 {chan1}]
+        assert_equal 2 [r publishlocal chan1 hello]
+        assert_equal "chan1 2" [r pubsub local numsub chan1]
+        assert_equal "chan1" [r pubsub local channels]
+
+        # clean up clients
+        $rd1 close
+        $rd2 close
+    }
+
+    test "PUBLISHLOCAL/SUBSCRIBELOCAL with PUBLISH/SUBSCRIBE" {
+        set rd1 [redis_deferring_client]
+        set rd2 [redis_deferring_client]
+
+        assert_equal {1} [subscribelocal $rd1 {chan1}]
+        assert_equal {1} [subscribe $rd2 {chan1}]
+        assert_equal 1 [r publishlocal chan1 hello]
+        assert_equal 1 [r publish chan1 hello]
+        assert_equal "chan1 1" [r pubsub local numsub chan1]
+        assert_equal "chan1 1" [r pubsub numsub chan1]
+        assert_equal "chan1" [r pubsub local channels]
+        assert_equal "chan1" [r pubsub channels]
+    }
+}
+
+start_server {tags {"pubsublocal"}} {
+start_server {} {
+    set node_0 [srv 0 client]
+    set node_0_host [srv 0 host]
+    set node_0_port [srv 0 port]
+    set node_0_pid [srv 0 pid]
+
+    set node_1 [srv -1 client]
+    set node_1_host [srv -1 host]
+    set node_1_port [srv -1 port]
+    set node_1_pid [srv -1 pid]
+
+    test {setup replication for following tests} {
+        $node_1 replicaof $node_0_host $node_0_port
+        wait_for_sync $node_1
+    }
+
+    test {publish message to replica and receive on master} {
```

Comment:
I don't understand how this test is working. Publishing to a replica in cluster mode disabled shouldn't allow it to be read on the master.

Reply:
You're right, it's incorrect. Strangely the test is succeeding and the test behaviour is also same for pubsub global semantics. Manually tested for CMD, publish to replica doesn't propagate to master.

 I've removed it, will try to figure out later why it's succeeding.